### PR TITLE
CRM-19463 Get E2E tests working on php7

### PR DESCRIPTION
**Pull request by seamuslee001**
@totten I would appreciate your thoughts on this, This feels awfully strange way of solving a problem i ran into running E2E tests locally on php7.

---
- [CRM-19463: Get E2E_AllTests working on php7](https://issues.civicrm.org/jira/browse/CRM-19463)

### CODE REVIEW

**Comment by totten** [comment]
The PR seems interesting in demonstrating where the problem lies,  but it's a hard compatibility break.  :(

**Comment by seamuslee001** [comment]
@totten Yeh Tim, I'm not sure mate This solved the problem for me, Now arguably i think we should be trying to keep it as an array in the SOAP and just ensure its encoded utf-8 but i couldn't figure it out and I don't understand SOAP enough

**Comment by seamuslee001** [comment]
@totten after some more googling i have come up with this and chucking some debugging in suggests that we are outputting in the same format from php5 to php7

**Comment by seamuslee001** [comment]
Jenkins re test this please

File: CRM/Utils/Array.php [hunk 1]
```diff
@@ -1076,4 +1076,23 @@ public static function convertCheckboxFormatToArray($input) {
...
+   *
+   * @return array $array utf8-encoded.
+   */
+  public static function encode_items($array) {
+    foreach ($array as $key => $value) {
+      if (is_array($value)) {
+        $array[$key] = self::encode_items($value);
+      }
+      else {
+        $array[$key] = mb_convert_encoding($value, mb_detect_encoding($value, mb_detect_order(), TRUE), 'UTF-8');
```
Comment (totten) [inline CRM/Utils/Array.php]: OK, this has the effect of changing how some return values are encoded in SOAP XML -- ie as `xsd:int` vs `xsd:string`. A quick fix seems to be:

```
diff --git a/CRM/Utils/Array.php b/CRM/Utils/Array.php
index 8537409..eaca889 100644
--- a/CRM/Utils/Array.php
+++ b/CRM/Utils/Array.php
@@ -1088,6 +1088,9 @@ class CRM_Utils_Array {
       if (is_array($value)) {
         $array[$key] = self::encode_items($value);
       }
+      elseif (is_int($value)) {
+        $array[$key] = $value;
+      }
       else {
         $array[$key] = mb_convert_encoding($value, mb_detect_encoding($value, mb_detect_order(), TRUE), 'UTF-8');
       }
```

Comment (totten) [inline CRM/Utils/Array.php]: Actually, there might be other primitive types. This seems better and also passes for me in PHP7:

```
diff --git a/CRM/Utils/Array.php b/CRM/Utils/Array.php
index 8537409..d82250c 100644
--- a/CRM/Utils/Array.php
+++ b/CRM/Utils/Array.php
@@ -1088,9 +1088,12 @@ class CRM_Utils_Array {
       if (is_array($value)) {
         $array[$key] = self::encode_items($value);
       }
-      else {
+      elseif (is_string($value)) {
         $array[$key] = mb_convert_encoding($value, mb_detect_encoding($value, mb_detect_order(), TRUE), 'UTF-8');
       }
+      else {
+        $array[$key] = $value;
+      }
     }
     return $array;
   }
```

**Comment by seamuslee001** [comment]
jenkins re test this please

**Comment by totten** [comment]
Tests passing. When I manually compared the XML outputted (http://php.net/manual/en/soapclient.getlastresponse.php) before and after the patch, it appeared equivalent.

**Pull request merged**